### PR TITLE
Metrics analyzer for Razor: Lines of code are outside the range of the file

### LINKED ISSUE #9288: Metrics analyzer for Razor: Lines of code are outside the range of the file
Community report:
https://community.sonarsource.com/t/java-lang-illegalargumentexception-line-575-is-out-of-range-for-file-xxx-file-has-386-lines/108740/15

The razor source generator of .Net SDK 9.0.100-preview.3.24204.13 produces a file with missing `#line default` and `#line hidden` directives for `using` directives:
```cs
                #line (1,2)-(2,1) "File.razor"
                using System.Net.Http

                #nullable disable
                    ;
                #nullable restore
```
Instead, the generated code should look like so:
```cs
                #line (1,2)-(2,1) "File.razor"
                using System.Net.Http

                #line default
                #line hidden
                #nullable disable
                    ;
                #nullable restore
```
As a result, we count the `;` as code line but the mapping is beyond the length of the file if the `using` is the only content of the file like in `_Imports.razor`

### PR DESCRIPTION
**Pull request by costin-zaharia-sonarsource**
Fixes #9288

For performance reasons, I want to avoid counting the file lines in the utility analyzers. To reduce the IO operations during analysis. An additional reason is that any information we get about invalid data will be lost during analysis (we don't have logs).

The idea is to do a check on the plugin side to see if the data we send to the server is valid. If it's not, we will log a debug message with the file name and the line number.

### CODE REVIEW

File: sonar-dotnet-shared-library/src/test/resources/RazorProtobufImporter/log.pb [hunk 1]
```diff
@@ -1 +1 @@
-Roslyn version: 4.8.0.0Language version: CSharp12!Concurrent execution: enabled��File 'Microsoft.NET.Sdk.Razor.SourceGenerators\Microsoft.NET.Sdk.Razor.SourceGenerators.RazorSourceGenerator\OverlapSymbolReferences_razor.g.cs' was recognized as razor generated��File 'Microsoft.NET.Sdk.Razor.SourceGenerators\Microsoft.NET.Sdk.Razor.SourceGenerators.RazorSourceGenerator\Cases_razor.g.cs' was recognized as razor generatedvrFile 'C:\Users\martin.strecker\Desktop\WebAss\obj\Debug\net8.0\WebAss.AssemblyInfo.cs' was recognized as generated��File 'C:\Users\martin.strecker\Desktop\WebAss\obj\Debug\net8.0\.NETCoreApp,Version=v8.0.AssemblyAttributes.cs' was recognized as generated
```
Comment (andrei-epure-sonarsource) [inline sonar-dotnet-shared-library/src/test/resources/RazorProtobufImporter/log.pb:1]: can we keep the old tests as well?

Comment (costin-zaharia-sonarsource) [reply sonar-dotnet-shared-library/src/test/resources/RazorProtobufImporter/log.pb:1]: to reflect the behavior for both versions of Roslyn?

Comment (andrei-epure-sonarsource) [reply sonar-dotnet-shared-library/src/test/resources/RazorProtobufImporter/log.pb:1]: yes

File: sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorImporterTestBase.java [hunk 2]
```diff
@@ -37,28 +37,26 @@
...
     return Paths.get(filePath).getFileName().toString();
   }
 
   @Before
-  public void setUp() throws FileNotFoundException {
+  public void setUp() {
     logTester.setLevel(Level.TRACE);
-    CasesInputFile = addTestFileToContext("Cases.razor");
-    OverlapSymbolReferencesInputFile = addTestFileToContext("OverlapSymbolReferences.razor");
-    ProgramInputFile = addTestFileToContext("Program.cs");
```
Comment (costin-zaharia-sonarsource) [inline sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorImporterTestBase.java:57]: Each test will add its own files to the context.

File: sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorMetricsImporterTest.java [hunk 3]
```diff
@@ -74,25 +81,73 @@ public void roslyn_metrics_are_imported() {
...
     verify(noSonarFilter).noSonarInFile(inputFile, Collections.emptySet());
 
-    verify(fileLinesContext).setIntValue(CoreMetrics.EXECUTABLE_LINES_DATA_KEY, 3, 1);
-    verify(fileLinesContext).setIntValue(CoreMetrics.EXECUTABLE_LINES_DATA_KEY, 5, 1);
-    verify(fileLinesContext).setIntValue(CoreMetrics.EXECUTABLE_LINES_DATA_KEY, 8, 1);
-    verify(fileLinesContext).setIntValue(CoreMetrics.EXECUTABLE_LINES_DATA_KEY, 13, 1);
-    verify(fileLinesContext).setIntValue(CoreMetrics.EXECUTABLE_LINES_DATA_KEY, 23, 1);
-    verify(fileLinesContext).setIntValue(CoreMetrics.EXECUTABLE_LINES_DATA_KEY, 24, 1);
-
-    verify(fileLinesContext, times(2)).setIntValue(CoreMetrics.NCLOC_DATA_KEY, 3, 1);
```
Comment (costin-zaharia-sonarsource) [inline sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorMetricsImporterTest.java:84]: The initial test was misleading. It added three files to the context and computed the metrics for all of them. At the same time, only one file was declared at the top as `inputFile`.

File: sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorMetricsImporterTest.java [hunk 4]
```diff
@@ -74,25 +81,73 @@ public void roslyn_metrics_are_imported() {
...
+        Tuple.tuple(CoreMetrics.COMMENT_LINES, 0),
+        Tuple.tuple(CoreMetrics.COGNITIVE_COMPLEXITY, 1),
+        Tuple.tuple(CoreMetrics.CLASSES, 0),
+        Tuple.tuple(CoreMetrics.NCLOC, 14),
+        Tuple.tuple(CoreMetrics.STATEMENTS, 3));
+
+    verify(noSonarFilter).noSonarInFile(inputFile, Collections.emptySet());
+
+    verifyMetrics(fileLinesContext, CoreMetrics.EXECUTABLE_LINES_DATA_KEY, 8, 23, 24);
+    verifyMetrics(fileLinesContext, CoreMetrics.NCLOC_DATA_KEY, 1, 3, 5, 8, 9, 13, 16, 18, 19, 21, 22, 23, 24, 25);
```
Comment (gregory-paidis-sonarsource) [inline sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorMetricsImporterTest.java:116]: Why does this include `1` on the non-comment lines here?
Line#1 is the `@page` directive. 
Is this just roslyn internal change or should we care about this?

File: sonar-dotnet-shared-library/src/main/java/org/sonarsource/dotnet/shared/plugins/protobuf/MetricsImporter.java [hunk 5]
```diff
@@ -62,8 +66,13 @@ void consumeFor(InputFile inputFile, MetricsInfo message) {
 
     saveMetric(context, inputFile, CoreMetrics.COMMENT_LINES, message.getNonBlankCommentCount());
 
+    int lineCount = inputFile.lines();
     for (int line : message.getCodeLineList()) {
-      fileLinesContext.setIntValue(CoreMetrics.NCLOC_DATA_KEY, line, 1);
+      if (line <= lineCount) {
+        fileLinesContext.setIntValue(CoreMetrics.NCLOC_DATA_KEY, line, 1);
+      } else if (LOG.isDebugEnabled()) {
```
[the remark this answers is not in the document]
Comment (gregory-paidis-sonarsource) [reply sonar-dotnet-shared-library/src/main/java/org/sonarsource/dotnet/shared/plugins/protobuf/MetricsImporter.java:73]: More info on this: 
If you remove the if and run `roslyn_metrics_out_of_range_with_4_10_debug_disabled`, it still passes.

File: sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorMetricsImporterTest.java [hunk 6]
```diff
@@ -74,25 +81,73 @@ public void roslyn_metrics_are_imported() {
...
+    assertThat(logTester.logs(Level.DEBUG)).isEmpty();
+  }
+
+  private void verifyMetrics(FileLinesContext context, String key, int... values) {
+    var groups = Arrays.stream(values)
+      .boxed()
+      .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+
+    for (int groupKey : groups.keySet()){
+      verify(context, times(groups.get(groupKey).intValue())).setIntValue(key, groupKey, 1);
```
Comment (gregory-paidis-sonarsource) [inline sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorMetricsImporterTest.java:150]: I find it a bit hard to understand what you are doing here.
Why do you group the lines and then do `times`?
From debugging they are always the unique.
How could it be that a group has more than one occurrence of each line?

Do you want to make sure exactly this? That every line appears ONLY once?

**Review by gregory-paidis-sonarsource — COMMENTED** [review]
Hey dude, this looks very nice and well-tested.
I left a couple of questions, mostly for curious/educational reasons.
I don't assume many changes will be needed, after my questions it will probably be approved. :)

File: sonar-dotnet-shared-library/src/main/java/org/sonarsource/dotnet/shared/plugins/protobuf/MetricsImporter.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (costin-zaharia-sonarsource) [reply sonar-dotnet-shared-library/src/main/java/org/sonarsource/dotnet/shared/plugins/protobuf/MetricsImporter.java:73]: Good question. `LOG.isDebugEnabled()` is recommended in situations like this one, when methods need to be invoked or code needs to be evaluated for logging purposes.

In this case:
```java
LOG.debug("The code line number was out of the range. File {}, Line {}", inputFile.filename(), line);
```
it's about `inputFile.filename()`. We have a rule that checks this behavior and raises if `isDebugEnabled` is not called.

[the remark this answers is not in the document]
Comment (gregory-paidis-sonarsource) [reply sonar-dotnet-shared-library/src/main/java/org/sonarsource/dotnet/shared/plugins/protobuf/MetricsImporter.java:73]: Good to know!

File: sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorMetricsImporterTest.java (hunk 4 above)
Comment (costin-zaharia-sonarsource) [reply sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorMetricsImporterTest.java:116]: Good catch! I think that's not correct and that it's caused by the generated code that has this snipped:
```
#nullable restore
#line (1,7)-(1,15) "\WebProject\Cases.razor"
"/cases"
```
I'll need to look into it and report an issue.

Comment (costin-zaharia-sonarsource) [reply sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorMetricsImporterTest.java:116]: I think in this PR we cannot handle that.

File: sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorMetricsImporterTest.java (hunk 6 above)
Comment (costin-zaharia-sonarsource) [reply sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorMetricsImporterTest.java:150]: When multiple files with test cases are added to the context, we can have lines reported multiple times. This was the behavior before my changes when the test base class added files to the context.

Strictly speaking, this is now used only to check that each element is added only once, and the query can be simplified. I wanted to have this method able to check duplicates, so we can easily add test cases in the future.

File: sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorMetricsImporterTest.java (hunk 4 above)
Comment (gregory-paidis-sonarsource) [reply sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorMetricsImporterTest.java:116]: Sure, no problem on my side. 
Resolving.

File: sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorMetricsImporterTest.java (hunk 6 above)
Comment (gregory-paidis-sonarsource) [reply sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorMetricsImporterTest.java:150]: That makes sense, thanks!

**Review by gregory-paidis-sonarsource — APPROVED** [review]
LGTM, thanks for the explanations!

File: sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorMetricsImporterTest.java (hunk 4 above)
Comment (costin-zaharia-sonarsource) [reply sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/plugins/protobuf/RazorMetricsImporterTest.java:116]: See: https://github.com/dotnet/razor/issues/10375#issuecomment-2124657614

**Pull request merged**